src/infrastructure/arranging/ordering.rs

```diff
@@ -0,0 +1,441 @@
+/// 1. `incoming_index` == `expected_index`.
+/// This package meets the expected order, so we can return it immediately.
+/// 2. `incoming_index` > `expected_index`.
+/// This package is newer than we expect, so we have to hold it temporarily until we have received all previous packages.
+/// 3. `incoming_index`< `expected_index`
+/// This can only happen in cases where we have a duplicated package. Again we don't give anything back.
+/// # Remarks
+/// - See [super-module](../index.html) for more information about streams.
+pub struct OrderingStream<T> {
+    // the id of this stream.
+    stream_id: u8,
+    // the storage for items that are waiting for older items to arrive.
+    // the items will be stored by key and value where the key is de incoming index and value the item value.
+    storage: HashMap<usize, T>,
+    // the next expected item index.
+    expected_index: usize,
+}
+
+impl<T> OrderingStream<T> {
+    /// Constructs a new, empty [`OrderingStream<T>`](./struct.OrderingStream.html).
+    ///
+    /// The default stream will have a capacity of 32 items.
+    pub fn new(stream_id: u8) -> OrderingStream<T> {
+        OrderingStream::with_capacity(1024, stream_id)
+    }
+
+    /// Constructs a new, empty [`OrderingStream`] with the specified capacity.
+    ///
+    /// The stream will be able to hold exactly capacity elements without
+    /// reallocating. If capacity is 0, the vector will not allocate.
+    ///
+    /// It is important to note that although the returned stream has the capacity specified,
+    /// the stream will have a zero length.
+    ///
+    /// [`OrderingStream`]: ./struct.OrderingStream.html
+    pub fn with_capacity(size: usize, stream_id: u8) -> OrderingStream<T> {
+        OrderingStream {
+            storage: HashMap::with_capacity(size),
+            expected_index: 1,
+            stream_id,
+        }
+    }
+
+    /// Returns the identifier of this stream.
+    fn stream_id(&self) -> u8 {
+        self.stream_id
+    }
+
+    /// Returns the next expected index.
+    pub fn expected_index(&self) -> usize {
+        self.expected_index
+    }
+
+    /// Returns an iterator of stored items.
+    ///
+    /// # Algorithm for returning items from an Iterator.
+    ///
+    /// 1. See if there is an item matching our `expected_index`
+    /// 2. If there is return the `Some(item)`
+    ///    - Increase the `expected_index`
+    ///    - Start at '1'
+    /// 3. If there isn't return `None`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// let stream = OrderingStream::new();
+    ///
+    /// let iter = stream.iter_mut();
+    ///
+    /// while let Some(item) = iter.next() {
+    ///    // Items from iterator are in order.
+    /// }
+    /// ```
+    ///
+    /// # Remarks
+    /// - Iterator mutates the `expected_index`.
+    /// - You can't use this iterator for iterating trough all cached values.
+    pub fn iter_mut(&mut self) -> IterMut<T> {
```

Comment:
This should probably have a more descriptive name than `IterMut`

Reply:
I don't agree, 'iter' an 'iter_mut' are standeralized names in rust, checkout the rust collections. Take note that the type returned here implements an iterator.

Reply:
I was specifically referring to the type name. I guess it is common in rust to have an `IterMut` struct in the namespace representing the mutable iterator so I guess it's fine (even if I don't love it).